Include/internal/pycore_c_array.h

```diff
@@ -0,0 +1,35 @@
+#ifndef Py_INTERNAL_C_ARRAY_H
+#define Py_INTERNAL_C_ARRAY_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef Py_BUILD_CORE
+#  error "this header requires Py_BUILD_CORE define"
+#endif
+
+
+/* Utility for a number of growing arrays */
+
+typedef struct {
+    void **array;             /* pointer to the array */
+    int *allocated_entries;   /* pointer to the capacity of the array */
+    size_t item_size;         /* size of each element */
+    int initial_num_entries;  /* initial allocation size */
+} _Py_c_array_t;
+
+/* If idx is out of bouds:
+ * If arr->array is NULL, allocate arr->initial_num_entries slots.
+ * Otherwise, double its size.
+ *
+ * Return 0 if successful and -1 (with exception set) otherwise.
+ */
+int _Py_c_array_EnsureCapacity(_Py_c_array_t *c_array, int idx);
```

Comment:
```suggestion
int
_Py_c_array_EnsureCapacity(_Py_c_array_t *c_array, int idx);
```

Should we assume idx is n int or maybe a uint32_t or just a size_t? any possibility of unsafe downcasting?